### ISSUE TITLE
timestamp timelocks are not satisfied

### ISSUE THREAD

**Issue by LLFourn** [bitcoindevkit/bdk#642]
It looks like the logic of `finalize_psbt` only checks if height based timelocks are satisfied:

```rust
                  match desc.satisfy(
                        &mut tmp_input,
                        (
                            PsbtInputSatisfier::new(psbt, n),
                            After::new(current_height, false),
                            Older::new(current_height, create_height, false),
                        ),
                    ) {
```

In addition timestamp we should check satisfaction of timestamp based timelocks. I will fix this in `bdk_core` but I thought this was worth noting and fixing ahead of then.

**Comment by wszdexdrf** [comment]
Maybe I could work on this.

**Comment by LLFourn** [comment]
Sure. To do this strictly correctly you need to know the "median time past": https://github.com/bitcoin/bitcoin/blob/a4e066af8573dcefb11dff120e1c09e8cf7f40c2/src/chain.h#L290-L302

I think an implementation on latest block timestamp would be better than nothing though?

**Comment by LLFourn** [comment]
Maybe relevant: https://github.com/rust-bitcoin/rust-miniscript/pull/408

**Comment by wszdexdrf** [comment]
I have reached a small hiccup. I was going to get blocktimes (for median time past) from the `blockchain` object, but then realised there is no way of actually accessing the blockchain inside `finalize_psbt` (no reference passed as an argument). What should we do here?

**Comment by afilini** [comment]
Wait, I'm a bit lost on this: the PR says that this is actually not a problem because miniscript correctly checks for the timestamps as well, so I guess we could close this issue, right?

Or did anybody ever encounter some actual issue trying to finalize a tx with timestamps?

**Comment by notmandatory** [comment]
Added to alpha release, need to confirm miniscript correctly checks for the timestamps and if so close. Otherwise need to see if/how to fix.

**Comment by nondiremanuel** [comment]
Need a check to see if it remains in alpha @notmandatory

**Comment by notmandatory** [comment]
@ValuedMammal per your note in discord. I found a "threshold" test that includes timelocks in the `rust-miniscript` plan module, but we're not yet using this. It's possible we don't have any tests right now that confirm timestamp timelocks are satisfied. But it should be possible to create a test to confirm it. Here's the plan module test: https://github.com/rust-bitcoin/rust-miniscript/blob/45904881e0d6f27f8b81ab310c4739cc14429a9b/src/plan.rs#L863

**Comment by f3r10** [comment]
hi @notmandatory is this issue still open? Could I give it a try?

**Comment by notmandatory** [comment]
Hi @f3r10, yes I don't think anyone else is working on it right now. To start I think you just need to create a unit test to verify if this is even an issue still as mentioned above. Thanks for taking a look.

**Comment by LLFourn** [comment]
FWIW I think this is a difficult issue to approach since I think it amounts to using the planning module correctly at this point. I'm not sure who's working on that if anyone.

**Comment by LLFourn** [comment]
It looks like your test passed @f3r10. So this means the `policy.rs` module already handled the constraints. Nice! Do we close this then?

**Comment by f3r10** [comment]
Yes @LLFourn the test is passing :tada:  Not sure yet if it would be necessary to add more tests so we could be completely sure :thinking:

**Comment by LLFourn** [comment]
@f3r10 When I look at the code I still see:


https://github.com/bitcoindevkit/bdk/blob/f6c1c61e7611966e20a88aace932a08aabd543f2/crates/wallet/src/wallet/mod.rs#L1877-L1886

I can't understand how time (not height) based locks actually work since we don't even tell the descriptor satisfier about the current timestamp?

**Comment by f3r10** [comment]
Yes, I can confirm that this is still an issue @LLFourn @notmandatory.
As you mention @LLFourn it would be necessary to tell to the `Satisfier` about the current timestamp, something like this:
```rust 
fn check_after(&self, n: u32) -> bool {
        if n < BLOCKS_TIMELOCK_THRESHOLD {
            if let Some(current_height) = self.current_height {
                current_height >= n
            } else {
                self.assume_reached
            }
        } else if let Some(current_time) = self.current_time {
            current_time >= n
        } else {
            self.assume_reached
        }
    }
 ```

**Comment by ValuedMammal** [comment]
Apparently excluding `After` and `Older` from the arguments doesn't impact the result (at least what we have test coverage for) so the logic seems to be contained in `PsbtInputSatisfier` I guess by comparing the locktime value we put in the PSBT with the value in the miniscript. That explains why we're not considering actual clock time here (not sure if that's intended). So if the descriptor contains a time-based absolute timelock, my thinking is that the user will have to tell us what timestamp to put in the PSBT.

**Comment by ValuedMammal** [comment]
> After further check, I found that the test I created is not passing. I only added this check which returns true `wallet.sign(&mut psbt, SignOptions::default()).unwrap();`
> 
> but calling directly `wallet.finalize_psbt(&mut psbt, SignOptions::default()).unwrap();` returns false 🤔
> 
> I am going to keep trying to understand the issue and I will let you know what I find @LLFourn

In either case the PSBT needs to be signed before it can be finalized. Calling `Wallet::sign` with the default `SignOptions` will attempt to do both.

**Comment by f3r10** [comment]
yes you are right @ValuedMammal  `Wallet::sign` with the default `SignOptions` calls  `Wallet::finalize_psbt` that returns true for the test I created.
```rust       
 // attempt to finalize
        if sign_options.try_finalize {
            self.finalize_psbt(psbt, sign_options)
        } else {
            Ok(false)
        }
```
But calling directly `Wallet::finalize_psbt` returns  `CouldNotSatisfy`.
I do not understand why :shrug:

**Comment by f3r10** [comment]
oh, I understand now, as @ValuedMammal said  `After and Older` are not used.  Only the first one `PsbtInputSatisfier::new(psbt, n)` is used, and `PsbtInputSatisfier` only checks if PSBT is signed correctly which is why `Wallet::sign` returns true `Wallet::finalize_psbt` not. 

The problem is that from arguments of the tuple  
```rust
( 
                 PsbtInputSatisfier::new(psbt, n), 
                 After::new(Some(current_height), false), 
                 Older::new(Some(current_height), confirmation_height, false), 
)
```
only the first one is used. If I commented `PsbtInputSatisfier::new(psbt, n)` then `After::new(Some(current_height), false)` is used and then the result is ` CouldNotSatisfy` with both `Wallet::sign` and `Wallet::finalize_psbt`

Hopefully I could explained :sweat_smile:

**Comment by evanlinjin** [comment]
In regards to testing this.

https://github.com/bitcoindevkit/bdk/pull/1670/files#r1873027475

This is related. We should have a test which fails if prev confirmation heights do not satisfy the timelock.
